File: src/backend/tracking-service/internal/models/location.go
// Package models provides data models for the tracking service
package models

import (
	"time"
)

// Human Tasks:
// 1. Ensure proper indexing for location data in the database
// 2. Configure monitoring for location data validation
// 3. Set up alerts for location tracking anomalies
// 4. Review and adjust location validation rules based on business requirements

// Location represents a geographical location with latitude, longitude, and timestamp.
// Addresses requirement: Technical Specification/7.2.1 Core Components/Tracking Service
// The Location model is used for real-time location tracking and processing.
type Location struct {
	// Latitude represents the geographical latitude coordinate
	Latitude float64 `json:"latitude" bson:"latitude"`

	// Longitude represents the geographical longitude coordinate
	Longitude float64 `json:"longitude" bson:"longitude"`

	// Timestamp represents when this location was recorded
	Timestamp time.Time `json:"timestamp" bson:"timestamp"`
}

// NewLocation creates a new Location instance with the provided coordinates and timestamp.
// Addresses requirement: Technical Specification/7.2.1 Core Components/Tracking Service
func NewLocation(latitude, longitude float64, timestamp time.Time) *Location {
	return &Location{
		Latitude:  latitude,
		Longitude: longitude,
		Timestamp: timestamp,
	}
}

// Validate performs validation checks on the Location instance.
// Addresses requirement: Technical Specification/7.2.1 Core Components/Tracking Service
func (l *Location) Validate() error {
	// Validate latitude range (-90 to 90)
	if l.Latitude < -90 || l.Latitude > 90 {
		return fmt.Errorf("invalid latitude: must be between -90 and 90")
	}

	// Validate longitude range (-180 to 180)
	if l.Longitude < -180 || l.Longitude > 180 {
		return fmt.Errorf("invalid longitude: must be between -180 and 180")
	}

	// Validate timestamp is not zero
	if l.Timestamp.IsZero() {
		return fmt.Errorf("invalid timestamp: cannot be zero")
	}

	// Validate timestamp is not in the future
	if l.Timestamp.After(time.Now()) {
		return fmt.Errorf("invalid timestamp: cannot be in the future")
	}

	return nil
}

File: src/backend/shared/models/location.ts
// class-validator v0.13.2
import { IsNumber, IsString, IsLatitude, IsLongitude, validate } from 'class-validator';
import logger from '../../shared/utils/logger';
import { createHttpError } from '../../shared/utils/error';

/**
 * Represents a geographical location with latitude, longitude, and address.
 * Addresses requirement: Technical Specification/7.2.1 Core Components/Tracking Service
 * Handles real-time location processing for tracking dog walks.
 */
export class Location {
  @IsNumber()
  @IsLatitude()
  public latitude: number;

  @IsNumber()
  @IsLongitude()
  public longitude: number;

  @IsString()
  public address: string;

  /**
   * Initializes a Location instance with latitude, longitude, and address.
   * @param latitude - The geographical latitude coordinate
   * @param longitude - The geographical longitude coordinate
   * @param address - The human-readable address of the location
   */
  constructor(latitude: number, longitude: number, address: string) {
    this.latitude = latitude;
    this.longitude = longitude;
    this.address = address;
  }
}

/**
 * Validates a Location instance using class-validator decorators.
 * Addresses requirement: Technical Specification/7.2.1 Core Components/Tracking Service
 * Ensures location data integrity for real-time tracking.
 * 
 * @param location - The Location instance to validate
 * @returns A promise resolving to an array of validation errors, if any
 * @throws HttpError if validation fails
 */
export async function validateLocation(location: Location): Promise<any[]> {
  try {
    const validationErrors = await validate(location);
    
    if (validationErrors.length > 0) {
      logger.logError('Location validation failed', {
        errors: validationErrors,
        location: {
          latitude: location.latitude,
          longitude: location.longitude,
          address: location.address
        }
      });
      
      throw createHttpError(400, 'Invalid location data provided');
    }
    
    return validationErrors;
  } catch (error) {
    logger.logError('Error during location validation', { error });
    throw error;
  }
}